README
		FEHASHMAC Version 2.1   07.09.2016

		Download and Verify

The latest and all earlier releases can be downloaded from
http://fehashmac.sourceforge.net
See also: https://www.hvf.ch/fehashmac/README.html

To verify the download, get the hash file fehashmac-n.n.n.tar.gz.hash
and execute the command
 $ fehashmac -c fehashmac-n.n.n.tar.gz.hash
This will test all algorithms, HMACs, KMACs and Base64.

		Documentation

See README.html in the Doc directory

New Features
------------

*07.09.2016: Replace SHA3, keccak with new code base*

*02.08.2016: General cleanup*
            One potential buffer overrun in keccak has been fixed.
            The code has been checked against STACK for null pointers,
            three finding have been fixed. Doc has been reorganized.
            A shell script, md5sum2md5 has been added to convert 
            from md5sum syntax to md5 style.

*11.07.2016: MD6 algorithms have been added*
            md6-224, md6-256, md6-384, md6-512, 
            although they had been retired from the SHA3 competition. 
            Performance comparisons are available in the Doc directory.

*29.02.2016: SHA3 algorithms are now based on FIPS 202 issued August 2015.*
            shake128, shake256 support now extendable length output,
            the previously defined algos xofshake128, xofshake256 are
            now superseded by shake128, shake256 and are therefore abandoned.

*20.05.2015: KMAC implemented for SHA3 algorithms and XOFs.*
            README updated. Algorithm kmac-all selects all algorithms
            that support KMAC, all-b64 selects all algos that support
            Base64 encoding, and similarly kmac-all-b64. The KMAC key can
            either be an ASCII string (up to 253 bytes long, --key=xxx) or a 
            hex string (--hexkey=nnn), whose length in bits may be specified
            with --keylength=bbb (max value 254*8-1= 2031 bits).

*01.05.2015: Extendable Output Functions (XOFs) xofshake128, xofshake256.*
            Output length can be specified to an arbitrary number of
            bytes up to 9 Exabytes or infinity. BASE64 output is OK.

*27.02.2015: BASE64 support for SHAKE.*
            SHAKE can now produce BASE64 output.

*18.01.2015: SHA3 performance:*  +
            SHA3 is now implemented based on the 
            KeccakWidth1600Opt64LCu6 version in the 
            KeccakCodePackage (downloaded on 11.12.2014): 
            +$ git clone https://github.com/gvanas/KeccakCodePackage+
            This version gives good performance both
            on 32 and 64 bit Intel platforms as well as on an
            AMD GEODE based appliance. Reports about other platforms
            like SPARC are welcome.

*02.01.2015: The 6 SHA3 algorithms are now supported,*  +
            based on the
            Draft FIPS Pub 202 (May 2014): sha3-224, sha3-256,
            sha3-384, sha3-512 and the two extendable output functions
            (XOF) shake128 and shake256. Test vectors are included.
            Our implementation is based on the reference implementation,
            a more efficient implementation will follow in an update.

*28.01.2012: Multiple algorithms (including HMACs) can be specified
            simultaneously.* +
            The data files are read only once, and all
            hashes are calculated in parallel.
            As an option, algorithm "all" expands to the list of all
            known algorithms, which are thus calculated in parallel for
            each data file or string. 
            Likewise, algorithm "hmac-all" expands to all known 
            HMAC algorithms.
            The results of the --check option and the test cases are 
            nicely summarized.

*29.03.2011: Support for the sha512-224 sha512-256 algorithms*

*14.04.2011: HMAC test vectors added* for MD5, SHA1, RMD128, RMD160.
            List of algorithms is now sorted alphabetically.

*13.04.2011: Support for all SHA-3 finalists:* BLAKE, GROESTL, JH, KECCAK,
            SKEIN for 224. 256. 384, 512 bits hash length, 
            SKEIN also for 1024 bits.
            They all support bitwise operation, bitwise test vectors are
            included (taken from the SHA-3 submissions).
            HMAC support upgraded to FIPS PUB 198-1 (2008),
            HMAC test vectors added for sha{224, 256, 384, 512}.

Purpose
-------

FEHASHMAC is a collection of publicly known hash algorithms integrated
          into a command-line utility. FEHASHMAC also contains a set
          of known test vectors and results for each algorithm
          such that the correct implementation for each hardware
          platform and compiler version can directly be verified.

FEHASHMAC supports bitwise hash calculation for algorithms with
          available bitwise test vectors. Currently this applies to
          the SHA algorithms: sha1, sha224, sha256, sha384, sha512,
          and to the SHA3 algorithms: sha3-224, sha3-256, sha3-384,
          sha3-512, shake128, shake256. The so-called Gillogly bitwise
          input has only been tested for sha1, but is also implemented
          in the SHA-2 hashes.

Bitwise hash calculation is also supported in sha512-224, sha512-256,
          whirl, but there are no bitwise test vectors available.

FEHASHMAC can also calculate hashed message authentication codes (HMAC)
          as specified in RFC 2104 and extended to arbitrary-length
          keys in FIPS PUB 198-1. The HMAC key can be specified as an
          ASCII string or as a hex string, and HMACs can be calculated
          for files, strings, and hex strings. 
          For SHA3, the KMAC message authentication is available, not HMAC.

To simplify usage, symbolic links with the algorithm name allow to call
fehashmac without the -a option. A prepended hmac- will automatically
enable HMAC mode.

Hashes and HMACs/KMACs of files can be verified with the --check option.

.Supported Algorithms (52) in 14 families:
**************************************************************

blake224 blake256 blake384 blake512 

gost 

groestl224 groestl256 groestl384 groestl512 

jh224 jh256 jh384 jh512 

keccak224 keccak256 keccak384 keccak512 

lash160 lash256 lash384 lash512 

md2 md4 md5 md6-224 md6-256 md6-384 md6-512

rmd128 rmd160 rmd256 rmd320 

sha1 

sha224 sha256 sha384 sha512 sha512-224 sha512-256 

sha3-224 sha3-256 sha3-384 sha3-512 shake128 shake256 

skein1024 skein224 skein256 skein384 skein512 

tiger2 

whirl 
**************************************************************

.Supported MAC Algorithms (46): 
**************************************************************

hmac-blake224 hmac-blake256 hmac-blake384 hmac-blake512 

hmac-gost 

hmac-groestl224 hmac-groestl256 hmac-groestl384 hmac-groestl512 

hmac-jh224 hmac-jh256 hmac-jh384 hmac-jh512 

hmac-keccak224 hmac-keccak256 hmac-keccak384 hmac-keccak512 

hmac-lash160 hmac-lash256 hmac-lash384 hmac-lash512 

hmac-md2 hmac-md4 hmac-md5 
         hmac-md6-224 hmac-md6-256 hmac-md6-384 hmac-md6-512

hmac-rmd128 hmac-rmd160 hmac-rmd256 hmac-rmd320 

hmac-sha1 

hmac-sha224 hmac-sha256 hmac-sha384 hmac-sha512 hmac-sha512-224 
            hmac-sha512-256 

hmac-skein224 hmac-skein256 hmac-skein384 hmac-skein512
            hmac-skein1024 

hmac-tiger2 

hmac-whirl 
**************************************************************

.Supported KMAC Algorithms (6): 
**************************************************************

kmac-sha3-224 kmac-sha3-256 kmac-sha3-384 kmac-sha3-512 
              kmac-shake128 kmac-shake256
**************************************************************

.Algorithms with Base64 Encoding (2): 
**************************************************************

shake128-b64 shake256-b64 
**************************************************************

.KMAC Algorithms with Base64 Encoding (2): 
**************************************************************

kmac-shake128-b64 kmac-shake256-b64 

**************************************************************

.Extendable Output Length (XOF) Algorithms (2):
**************************************************************

shake128 shake256

**************************************************************

SHA3 supports KMAC, not HMAC



NEWS
07.09.2016 Release 2.1
The SHA3 and keccak code base has been replaced by a current version
taken on 2016-08-04 from the github repository, 
https://github.com/gvanas/KeccakCodePackage.
A patch has been added to KeccakHash.c to correctly treat partial byte input.

11.07.2016 Release 1.5.1
The MD6 algorithms have been added - not that they are of practical use,
but now the complete set of Ron Rivest's hash codes (md4, md5, md6-*) 
is available.
Performance tests can now be executed in the Doc directory 
(type ``make perftest´´). It has been found that compiling with gcc and clang
gives different but inconsistent results. Results are included.

29.02.2015 Release 1.5.0
SHA3 is now based on the official FIPS 202 standard.
shake128 and shake256 support now the extendable length output,
the previous xofshake128 and xofshake256 algorithms are abandoned.

20.05.2015 Release 1.4.2
KMAC implemented for all SHA3 algorithms. Key is either
an ASCII  string of up to 253 bytes or a hex string of
up to 2031 bits. --keylength=nnn specifies the length 
of the key in bits (hexkey only).

01.05.2015 Release 1.4.1
Extendable Output Functions xofshake128, xofshake256 can now
produce arbitrary length output, also in base64 encoding
if desired.

27.02.2015 Release 1.4.0
shake128, shake256 can now produce and verify base64-encoded
output, use either the --base64 flag or specifiy the
algorithms as shake128-b64 and shake256-b64.

18.01.2015 Release 1.3.1
implements the performant SHA3 algorithms, derived from
the 64 bit KeccakWidth1600Opt64LCu6 version in 
KeccakCodePackage (downloaded on 11.12.2015): 

$ git clone https://github.com/gvanas/KeccakCodePackage

This version gives good performance both
on 32 and 64 bit Intel platforms as well as on an
AMD GEODE based appliance. Reports about other platforms
like SPARC are welcome.

03.01.2015 Release 1.3.0
includes the DRAFT SHA3 algorithms: sha3-224, sha3-256,
sha3-384, sha3-512, shake128, shake256.
The implementation of SHA3 is based on the reference 
implementation of the Keccak authors. A more performant
implementation will follow in an update.



lib/md2.c
/* MD2C.C - RSA Data Security, Inc., MD2 message-digest algorithm
 */

/*
 * Generic Hash and HMAC Program
 *
 * Copyright (C) 2009 Harald von Fellenberg <hvf@hvf.ch>
 *
 * This program is free software; you can redistribute it and/or modify 
 * it under the terms of the GNU General Public License as published by 
 * the Free Software Foundation; either version 3 of the License, or 
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty of 
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License 
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

/* Copyright (C) 1990-2, RSA Data Security, Inc. Created 1990. All
   rights reserved.

   License to copy and use this software is granted for
   non-commercial Internet Privacy-Enhanced Mail provided that it is
   identified as the "RSA Data Security, Inc. MD2 Message Digest
   Algorithm" in all material mentioning or referencing this software
   or this function.

   RSA Data Security, Inc. makes no representations concerning either
   the merchantability of this software or the suitability of this
   software for any particular purpose. It is provided "as is"
   without express or implied warranty of any kind.

   These notices must be retained in any copies of any part of this
   documentation and/or software.
 */

/* integrated into fehashmac - hvf 12.08.01 
 * hvf 10.02.2009 adapt to SHA3-C-API 
 * disabled MD2_Print - hvf 19.04.2015
 */

#include	"mdx.h"
#include	<string.h>
#include	<stdlib.h>
#include	<stdio.h>

#define DEBUG
#undef DEBUG

static void MD2_Transform (BYTE [16], BYTE [16], const BYTE [16]);

/* Permutation of 0..255 constructed from the digits of pi. It gives a
   "random" nonlinear byte substitution operation.
 */
static BYTE PI_SUBST[256] = {
	41, 46, 67, 201, 162, 216, 124, 1, 61, 54, 84, 161, 236, 240, 6,
	19, 98, 167, 5, 243, 192, 199, 115, 140, 152, 147, 43, 217, 188,
	76, 130, 202, 30, 155, 87, 60, 253, 212, 224, 22, 103, 66, 111, 24,
	138, 23, 229, 18, 190, 78, 196, 214, 218, 158, 222, 73, 160, 251,
	245, 142, 187, 47, 238, 122, 169, 104, 121, 145, 21, 178, 7, 63,
	148, 194, 16, 137, 11, 34, 95, 33, 128, 127, 93, 154, 90, 144, 50,
	39, 53, 62, 204, 231, 191, 247, 151, 3, 255, 25, 48, 179, 72, 165,
	181, 209, 215, 94, 146, 42, 172, 86, 170, 198, 79, 184, 56, 210,
	150, 164, 125, 182, 118, 252, 107, 226, 156, 116, 4, 241, 69, 157,
	112, 89, 100, 113, 135, 32, 134, 91, 207, 101, 230, 45, 168, 2, 27,
	96, 37, 173, 174, 176, 185, 246, 28, 70, 97, 105, 52, 64, 126, 15,
	85, 71, 163, 35, 221, 81, 175, 58, 195, 92, 249, 206, 186, 197,
	234, 38, 44, 83, 13, 110, 133, 40, 132, 9, 211, 223, 205, 244, 65,
	129, 77, 82, 106, 220, 55, 200, 108, 193, 171, 250, 36, 225, 123,
	8, 12, 189, 177, 74, 120, 136, 149, 139, 227, 99, 232, 109, 233,
	203, 213, 254, 59, 0, 29, 57, 242, 239, 183, 14, 102, 88, 208, 228,
	166, 119, 114, 248, 235, 117, 75, 10, 49, 68, 80, 180, 143, 237,
	31, 26, 219, 153, 141, 51, 159, 17, 131, 20
};

static BYTE *PADDING[] = {
	(BYTE *)"",
	(BYTE *)"\001",
	(BYTE *)"\002\002",
	(BYTE *)"\003\003\003",
	(BYTE *)"\004\004\004\004",
	(BYTE *)"\005\005\005\005\005",
	(BYTE *)"\006\006\006\006\006\006",
	(BYTE *)"\007\007\007\007\007\007\007",
	(BYTE *)"\010\010\010\010\010\010\010\010",
	(BYTE *)"\011\011\011\011\011\011\011\011\011",
	(BYTE *)"\012\012\012\012\012\012\012\012\012\012",
	(BYTE *)"\013\013\013\013\013\013\013\013\013\013\013",
	(BYTE *)"\014\014\014\014\014\014\014\014\014\014\014\014",
	(BYTE *)
	"\015\015\015\015\015\015\015\015\015\015\015\015\015",
	(BYTE *)
	"\016\016\016\016\016\016\016\016\016\016\016\016\016\016",
	(BYTE *)
	"\017\017\017\017\017\017\017\017\017\017\017\017\017\017\017",
	(BYTE *)
	"\020\020\020\020\020\020\020\020\020\020\020\020\020\020\020\020"
};

void MD2_update_old (MD2_CTX *context, const BYTE *input, unsigned int inputLen);
void MD2_final_old (MD2_CTX *context);

/* MD2 initialization. Begins an MD2 operation, writing a new context.
 */

HashReturn MD2_init (hashState  *state, int hashbitlen)
{
	/* verify correct hash length	*/
	if (hashbitlen != HASH_BITLENGTH_MD2)
		return BAD_HASHBITLEN;

	/* allocate context and fill it	*/
	MD2_CTX *context = (MD2_CTX *)malloc (sizeof (MD2_CTX));
	memset (context, 0, sizeof (MD2_CTX));
	context->hashbitlen = HASH_BITLENGTH_MD2;
	context->magic = HASH_MAGIC_MD2;

	*state = (hashState *) context;
	return SUCCESS;
}

HashReturn 	MD2_update (
	hashState state, 			/* previously initialized context */
	const BitSequence *buffer, 	/* bit buffer, first bit is MSB in [0] */
	DataLength databitlen)		/* number of bits to process from buffer */
{	
	/* can be called once or many times */
	/* verify correct hashbitlen and magic	*/

	MD2_CTX *context = (MD2_CTX *) state;
	if (context->hashbitlen != HASH_BITLENGTH_MD2)
		return BAD_HASHBITLEN;

	if (context->magic != HASH_MAGIC_MD2)
		return BAD_ALGORITHM;

	MD2_update_old (context, buffer, (unsigned int)(databitlen >> 3));

	return SUCCESS;
}

void MD2_update_old (MD2_CTX *context, const BYTE *input, unsigned int inputLen)
{	/* can be called once or many times	*/
	unsigned int i, index, partLen;

	/* Update number of bytes mod 16 */
	index = context->count;
	context->count = (index + inputLen) & 0xf;

	partLen = 16 - index;

	/* Transform as many times as possible.
	*/
	if (inputLen >= partLen) {
		memcpy ((BYTE *)&context->buffer[index], (BYTE *)input, partLen);
		MD2_Transform (context->state, context->checksum, context->buffer);

		for (i = partLen; i + 15 < inputLen; i += 16)
		  MD2_Transform (context->state, context->checksum, &input[i]);

		index = 0;
	}
	else
		i = 0;

	/* Buffer remaining input */
	memcpy ((BYTE *)&context->buffer[index], (BYTE *)&input[i], inputLen-i);
}

HashReturn	MD2_final (hashState state, BitSequence *hashval)
{	/*	does padding of last block	*/
	MD2_CTX *context = (MD2_CTX *) state;
	/* verify correct hashbitlen and magic	*/
	if (context->hashbitlen != HASH_BITLENGTH_MD2)
		return BAD_HASHBITLEN;

	if (context->magic != HASH_MAGIC_MD2)
		return BAD_ALGORITHM;

	MD2_final_old (context);

	if (hashval)
		memcpy (hashval, context->out, HASH_LENGTH_MD2);
	return SUCCESS;
}

void MD2_final_old (MD2_CTX *context)
{	/* does padding of last block	*/
	unsigned int index, padLen;

	/* Pad out to multiple of 16.
	*/
	index = context->count;
	padLen = 16 - index;
	MD2_update_old (context, PADDING[padLen], padLen);

	/* Extend with checksum */
	MD2_update_old (context, context->checksum, 16);

	/* Store state in digest */
	memcpy (context->out, context->state, 16);

}

/* MD2 basic transformation. Transforms state and updates checksum
     based on block.
 */
static void MD2_Transform (BYTE state[16], BYTE checksum[16], const BYTE block[16])
{
	unsigned int i, j, t;
	BYTE x[48];

#ifdef DEBUG
	for (i=0; i<16; i++)
		printf ("W[%d] = %2.2x\n", i, block[i]);
#endif

	/* Form encryption block from state, block, state ^ block.
	*/
	memcpy ((BYTE *)x, (BYTE *)state, 16);
	memcpy ((BYTE *)x+16, (BYTE *)block, 16);
	for (i = 0; i < 16; i++)
		x[i+32] = state[i] ^ block[i];

	/* Encrypt block (18 rounds).
	*/
	t = 0;
	for (i = 0; i < 18; i++) {
		for (j = 0; j < 48; j++)
			t = x[j] ^= PI_SUBST[t];
		t = (t + i) & 0xff;
	}

	/* Save new state */
	memcpy ((BYTE *)state, (BYTE *)x, 16);

	/* Update checksum.
	*/
	t = checksum[15];
	for (i = 0; i < 16; i++)
		t = checksum[i] ^= PI_SUBST[block[i] ^ t];

	/* Zeroize sensitive information.
	*/
	memset ((BYTE *)x, 0, sizeof (x));
}
/* utility routines
 * adapted for MDx hashes
 *
 * hvf 23.1.2000
 * hvf 13.2.01
 * hvf 10.8.01
 * hvf 12.8.01
 */

/* Digests a file and prints the result.
 */

HashReturn MD2_File (hashState state, FILE *in)
{
	MD2_CTX *context = (MD2_CTX *) state;
	int len;
	unsigned char buffer[BUFFERSIZE];
	HashReturn retval;

	/* verify correct hashbitlen and magic	*/
	if (context->hashbitlen != HASH_BITLENGTH_MD2)
		return BAD_HASHBITLEN;

	if (context->magic != HASH_MAGIC_MD2)
		return BAD_ALGORITHM;

	while ((len = fread (buffer, 1, BUFFERSIZE, in))) {
		retval = MD2_update (context, buffer, (DataLength)len << 3);
		if (retval != SUCCESS) return retval;
	}
	retval = MD2_final (context, NULL);

	fclose (in);
	return retval;
}

HashReturn MD2_HashToByte (hashState state, BYTE *out) 
{
	MD2_CTX *context = (MD2_CTX *) state;

	/* verify correct hashbitlen and magic	*/
	if (context->hashbitlen != HASH_BITLENGTH_MD2)
		return BAD_HASHBITLEN;

	if (context->magic != HASH_MAGIC_MD2)
		return BAD_ALGORITHM;

	memcpy (out, context->out, HASH_LENGTH_MD2);
	return SUCCESS;
}

/* all-in-one hash function as required by SHA3-C-API */

HashReturn MD2_hash (int hashbitlen, const BitSequence *data,
                      DataLength databitlen, BitSequence *hashval) 
{
	hashState	state;
	HashReturn	retval;
	
	retval = MD2_init (&state, HASH_BITLENGTH_MD2);
	//HashReturn MD2_init (hashState  *state, int hashbitlen)
	if (retval != SUCCESS) {
		fprintf (stderr, "MD2_init failed, reason %d, hash length %d\n",
                 retval, HASH_BITLENGTH_MD2);
        exit (1);
    }

	retval = MD2_update (state, data, databitlen);
	if (retval != SUCCESS) {
		fprintf (stderr, "MD2_update failed, reason %d\n", retval);
        exit (1);
    }

	retval = MD2_final (state, hashval);
	if (retval != SUCCESS) {
		fprintf (stderr, "MD2_final failed, reason %d\n", retval);
        exit (1);
    }
	free (state);
 	return retval;
}

